internal/olm/operator/internal/registry_pod.go

```diff
@@ -0,0 +1,293 @@
+	semverBundleAddMode      = "semver"
+	replacesBundleAddMode    = "replaces"
+)
+
+// RegistryPod holds resources necessary for creation of a registry server
+type RegistryPod struct {
+	// BundleAddMode specifies the graph update mode that defines how channel graphs are updated
+	BundleAddMode string
+
+	// BundleImage specifies the container image that opm uses to generate and incrementally update the database
+	BundleImage string
+
+	// Index image contains a database of pointers to operator manifest content that is queriable via an API.
+	// new version of an operator bundle when published can be added to an index image
+	IndexImage string
+
+	// DBPath refers to the registry DB;
+	// if an index image is provided, the existing registry DB is located at /database/index.db
+	DBPath string
+
+	// Namespace refers to the specific namespace in which the registry pod will be created and scoped to
+	Namespace string
+
+	// Kubeclient refers to a Kubernetes clientset that implements kubernetes.Interface.
+	Kubeclient kubernetes.Interface
+
+	// GRPCPort is the container grpc port which is defaulted to 50051
+	GRPCPort int32
+}
+
+// newRegistryPod initializes the RegistryPod struct and sets defaults for empty fields
+func newRegistryPod(kubeclient kubernetes.Interface, dbPath, bundleImage, namespace string) *RegistryPod {
+	rp := &RegistryPod{}
+
+	if rp.GRPCPort == 0 {
+		rp.GRPCPort = defaultGRPCPort
+	}
+
+	if rp.IndexImage == "" {
+		rp.IndexImage = defaultIndexImage
+	}
+
+	if rp.BundleAddMode == "" {
+		if rp.IndexImage == defaultIndexImage {
+			rp.BundleAddMode = semverBundleAddMode
+		} else {
+			rp.BundleAddMode = replacesBundleAddMode
+		}
+	}
+	rp.Kubeclient = kubeclient
+	rp.DBPath = dbPath
+	rp.BundleImage = bundleImage
+	rp.Namespace = namespace
+
+	return rp
+}
+
+// CreateOnCluster returns a bundle registry pod built from an index image
+// after verifying successful creation of the pod, or returns the pod logs and error in case of failures
+func (rp *RegistryPod) CreateOnCluster(ctx context.Context) (*corev1.Pod, string, error) {
+	// validate the RegistryPod struct and ensure required fields are set
+	err := rp.validate()
+	if err != nil {
+		return nil, "", fmt.Errorf("error in validating RegistryPod struct: %v", err)
+	}
+
+	// create the registry pod on the cluster
+	pod, err := rp.createPodOnCluster(ctx)
+	if err != nil {
+		podLogs, logErr := rp.FetchPodLogs(ctx, pod)
+		if logErr != nil {
+			return nil, podLogs, fmt.Errorf("error creating pod: %v: and fetching logs: %v", err, logErr)
+		}
+		return nil, podLogs, fmt.Errorf("error creating pod: %v", err)
+	}
+
+	// verify that pod is successfully created
+	err = rp.validatePodOnCluster(ctx, pod)
+	if err != nil {
+		podLogs, logErr := rp.FetchPodLogs(ctx, pod)
+		if logErr != nil {
+			return nil, podLogs, fmt.Errorf("error verifying pod creation: %v: and fetching logs: %v", err, logErr)
+		}
+		return nil, podLogs, fmt.Errorf("error verifying pod creation: %v", err)
+	}
```

Comment:
IMO the validation should not be made part of the create. Who will consume it can call create and then validate if and when wish. It is hurting the singles responsibility principle.

Reply:
so, if we need to return the pod logs when an error is faced it should be in the `RegistryPod.ValidateOnCluster`

Reply:
Create and VerifyPodRunning are two separate functions now. Create handles the creation part. When the CLI is plugged in, we will call `VerifyPodRunning` after calling `Create` to ensure that the pod status is running.

Reply:
Will we ever want to call `Create` without calling `VerifyPodRunning`? I don't think so, therefore it should be part of `Create`. We do something similar in [`Client.DoDelete()`](https://github.com/operator-framework/operator-sdk/blob/9907f30116a5ccdd8e43181f69700135ebcdca18/internal/olm/client/client.go#L98).

Reply:
Yes, I have the same opinion. It makes no sense to return a failed pod (for whatever reason), and the CLI then calls `VerifyPodRunning` to check the status. I think `Create` should only return after successfully verifying that it created a registry pod and after validating that its status is `Running`.

Reply:
Hi @estroz, 

A method called Create should just create and return success when the Pod is created. Otherwise, it can create the pod successfully and then return an error because it is not running. Create is one action, check if is running is another action and IMO not respect that means hurt concepts as single responsibility and cushion to say a few. 

However, if you both think that it makes more sense I do not think that it should be a blocker.